File: simulator/ts/geo/segment-line.ts
import { Point, PointV } from "../dtypes"
import { ParaControls } from "../paracontrols"
import { interpolate } from "../util"
import { Path } from "./paths"

/**
 * 2D line segment.
 */
export class SegmentLine {
  start: PointV
  end: PointV

  constructor(start: Point, end: Point) {
    const dx = end.x - start.x
    const dy = end.y - start.y
    const len = Math.hypot(dx, dy)
    this.start = {
      ...start,
      vx: dx / len,
      vy: dy / len
    }
    this.end = {
      ...end,
      vx: dx / len,
      vy: dy / len
    }
    if (len === 0) {
      console.error("Invalid line: same start and end", start)
    }
  }

  public controls(): ParaControls {
    return {
      left: 0,
      right: 0
    }
  }

  public length() {
    const dx = this.end.x - this.start.x
    const dy = this.end.y - this.start.y
    return Math.hypot(dx, dy)
  }

  /**
   * Fly a given distance along the path.
   * If the end of the path is reached, extend a straight line out to infinity.
   */
  public fly(distance: number): Path {
    if (distance <= 0) {
      console.error("line flight distance must be positive", distance)
    }
    // Linear interpolate
    const alpha = distance / this.length()
    return new Path("line-fly", new SegmentLine(this.start, {
      x: interpolate(this.start.x, this.end.x, alpha),
      y: interpolate(this.start.y, this.end.y, alpha)
    }))
  }

  public render(): Point[] {
    return [this.start, this.end]
  }
}

File: android/app/src/main/java/ws/baseline/autopilot/geo/SegmentLine.java
package ws.baseline.autopilot.geo;

import java.util.Arrays;
import java.util.List;

import static ws.baseline.autopilot.util.Numbers.interpolate;

public class SegmentLine implements Segment {
    private final Point start;
    private final Point end;

    public SegmentLine(Point start, Point end) {
        this.start = start;
        this.end = end;
    }

    @Override
    public Point start() {
        return start;
    }

    @Override
    public Point end() {
        return end;
    }

    @Override
    public PathLike fly(double distance) {
        // Linear interpolate
        final double alpha = distance / this.length();
        final Point proj = new Point(
                interpolate(this.start.x, this.end.x, alpha),
                interpolate(this.start.y, this.end.y, alpha)
        );
        return new SegmentLine(start, proj);
    }

    @Override
    public double length() {
        final double dx = start.x - end.x;
        final double dy = start.y - end.y;
        return Math.hypot(dx, dy);
    }

    @Override
    public List<Point> render() {
        return Arrays.asList(start, end);
    }
}
